### PR TITLE
[Merged by Bors] - feat(Algebra/Module): presentation of the cokernel of a linear map

### PR DESCRIPTION
**Pull request by joelriou**
If `f : M₁ →ₗ[A] M₂` is a linear map between modules, `pres₂` is a presentation of `M₂` and `g₁ : ι → M₁` is a family of generators of `M₁` (which is expressed as `hg₁ : Submodule.span A (Set.range g₁) = ⊤`), then we provide a way to obtain a presentation of the cokernel of `f`. It requires an additional data `data : pres₂.CokernelData f g₁`, which consists of liftings of the images by `f` of the generators of `M₁` as linear combinations of the generators of `M₂`. Then, we obtain a presentation `pres₂.cokernel data hg₁ : Presentation A (M₂ ⧸ LinearMap.range f)`.

---

- [x] depends on: #18295
- [x] depends on: #18327
- [x] depends on: #18408

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Exact.lean [hunk 1]
```diff
@@ -387,6 +392,40 @@ lemma Exact.exact_mapQ_iff {f : M →ₗ[R] N} {g : N →ₗ[R] P}
...
+  simp only [Submodule.ext_iff, mem_ker, Submodule.mem_bot, mem_range]
+  constructor
+  · intro hfg x
+    simpa using hfg (Submodule.Quotient.mk x)
+  · intro hfg x
+    obtain ⟨x, rfl⟩ := Submodule.Quotient.mk_surjective _ x
+    simpa using hfg x
+
+lemma injective_range_liftQ_of_exact (h : Function.Exact f g) :
+    Function.Injective ((range f).liftQ g (by simp only [h.linearMap_ker_eq, le_refl])) := by
```
Comment (erdOne) [inline Mathlib/Algebra/Exact.lean:416]:
```suggestion
    Function.Injective ((range f).liftQ g (h · |>.mpr)) := by
```

File: Mathlib/Algebra/Exact.lean [hunk 2]
```diff
@@ -387,6 +392,40 @@ lemma Exact.exact_mapQ_iff {f : M →ₗ[R] N} {g : N →ₗ[R] P}
...
+
+end LinearMap
+
+/-- The linear equivalence `(N ⧸ LinearMap.range f) ≃ₗ[A] P` associated to
+an exact sequence `M → N → P → 0` of `R`-modules. -/
+@[simps! apply]
+noncomputable def Function.Exact.linearEquivOfSurjective (h : Function.Exact f g)
+    (hg : Function.Surjective g) : (N ⧸ LinearMap.range f) ≃ₗ[R] P :=
+  LinearEquiv.ofBijective ((LinearMap.range f).liftQ g
+    (by simp only [h.linearMap_ker_eq, le_refl]))
```
Comment (erdOne) [inline Mathlib/Algebra/Exact.lean:427]:
```suggestion
  LinearEquiv.ofBijective ((LinearMap.range f).liftQ g (h · |>.mpr))
```

File: Mathlib/Algebra/Module/Presentation/Cokernel.lean [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+`g₁ : ι → M₁` of `M₁`, and an additional data in a `Presentation.CokernelData` structure. -/
+@[simps!]
+noncomputable def ofExact {f : M₁ →ₗ[A] M₂} {g : M₂ →ₗ[A] M₃}
+    (pres₂ : Presentation.{w₂₀, w₂₁} A M₂) {ι : Type w₁} {g₁ : ι → M₁}
+    (data : pres₂.CokernelData f g₁)
+    (hfg : Function.Exact f g) (hg : Function.Surjective g)
+    (hg₁ : Submodule.span A (Set.range g₁) = ⊤) :
+    Presentation A M₃ :=
+  ofIsPresentation ((pres₂.cokernel data hg₁).ofLinearEquiv
+    (hfg.linearEquivOfSurjective hg))
```
Comment (erdOne) [inline Mathlib/Algebra/Module/Presentation/Cokernel.lean:144]: This is `Presentation.ofLinearEquiv` from #18359 right?
Also `IsPresentation.ofLinearEquiv` should be `IsPresentation.of_linearEquiv`.

Comment (joelriou) [reply Mathlib/Algebra/Module/Presentation/Cokernel.lean]: Thanks very much! I have fixed these inconsistencies in these PRs.

**Comment by erdOne** [comment]
Thanks!
maintainer merge

**Comment by jcommelin** [comment]
bors merge

**Pull request closed without merging**